### PR TITLE
Fixed exception-controller to resolve parameters like websitecontroller

### LINKED ISSUE #1025: URLs variable missing in error page
If you request a not existent page the `urls`-variable is missing, and if any twig template is using it you'll get a 500 error instead of a 404. Should be added in the `ExceptionController` of the WebsiteBundle.

### PR DESCRIPTION
**Pull request by wachterjohannes**
__tasks:__

- [x] test coverage
- [ ] gather feedback for my changes
- [x] error template config in webspace

__related:__

* https://github.com/sulu-io/sulu-standard/pull/442

__informations:__

| q                | a
| ---------------- | ---
| Fixed tickets    | fixes #1025
| BC breaks        | see UPGRADE.md 
| Documentation PR | https://github.com/sulu-io/sulu-docs/pull/27

### CODE REVIEW

File: src/Sulu/Bundle/WebsiteBundle/Resolver/ParameterResolver.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function resolve(
+        array $parameter,
+        RequestAnalyzerInterface $requestAnalyzer = null,
+        StructureInterface $structure = null,
+        $preview = false
+    )
+    {
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Resolver/ParameterResolver.php]: Previous two lines should be one

File: src/Sulu/Bundle/WebsiteBundle/Controller/ExceptionController.php [hunk 2]
```diff
@@ -47,45 +73,30 @@ public function showAction(
...
+            $parameter = array(
+                'status_code' => $code,
+                'status_text' => isset(Response::$statusTexts[$code]) ? Response::$statusTexts[$code] : '',
+                'exception' => $exception,
+                'currentContent' => $currentContent
+            );
+            $data = $this->parameterResolver->resolve(
+                $parameter,
+                $this->requestAnalyzer,
+                $this->contentMapper->loadStartPage(
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Controller/ExceptionController.php]: For what data do you have to load the startpage here again?

File: src/Sulu/Bundle/WebsiteBundle/Resolver/ParameterResolver.php (hunk 1 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/WebsiteBundle/Resolver/ParameterResolver.php]: fixed

File: src/Sulu/Bundle/WebsiteBundle/Controller/ExceptionController.php (hunk 2 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/WebsiteBundle/Controller/ExceptionController.php]: i have removed that for a default `/` rlp

**Comment by wachterjohannes** [comment]
@danrot everything fixed!

File: src/Sulu/Bundle/WebsiteBundle/Controller/ExceptionController.php [hunk 3]
```diff
@@ -47,47 +73,35 @@ public function showAction(
...
-                $code = $exception->getStatusCode();
+            $currentContent = $this->getAndCleanOutputBuffering($request->headers->get('X-Php-Ob-Level', -1));
+            $code = $exception->getStatusCode();
+
+            // symfony exception template uses status_code and status_text
+            $parameter = array(
+                'statusCode' => $code,
+                'status_code' => $code,
+                'statusText' => isset(Response::$statusTexts[$code]) ? Response::$statusTexts[$code] : '',
+                'status_text' => isset(Response::$statusTexts[$code]) ? Response::$statusTexts[$code] : '',
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Controller/ExceptionController.php]: Shouldn't we agree on one of these formats? That screams for errors :confused:

Comment (wachterjohannes) [inline src/Sulu/Bundle/WebsiteBundle/Controller/ExceptionController.php]: @danrot the problem is that the symfony error template uses different variable name types(= for example `status_text` and `currentContent` ... we could kik the compatibility to symfony but it seems notgood for me!

Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Controller/ExceptionController.php]: My suggestion would have been to use the names from symfony... And maybe also use `current_content` then? /cc @chirimoya

Comment (wachterjohannes) [inline src/Sulu/Bundle/WebsiteBundle/Controller/ExceptionController.php]: but symfony uses also `currentContent` which is weird!

File: UPGRADE.md [hunk 4]
```diff
@@ -8,6 +8,7 @@ The `Sulu` prefix from all `ContentNavigationProviders` and `Admin` classes has
 been removed. You have to change these names in all usages of this classes in
 your own code.
 
+<<<<<<< HEAD
```
Comment (danrot) [inline UPGRADE.md]: Should not be here :)

File: src/Sulu/Bundle/WebsiteBundle/Controller/ExceptionController.php [hunk 5]
```diff
@@ -28,68 +29,65 @@ class ExceptionController extends BaseExceptionController
...
-                            )
-                        )
-                    ),
-                    $exception->getStatusCode()
-                );
-            }
-        }
+        $parameter = array(
+            'statusCode' => $code,
+            'statusText' => isset(Response::$statusTexts[$code]) ? Response::$statusTexts[$code] : '',
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Controller/ExceptionController.php]: Didn't we say we use the names from symfony, although they are inconsistent?

File: src/Sulu/Bundle/WebsiteBundle/Resolver/ParameterResolver.php [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            /** @var Localization $localization */
+            $locale = $localization->getLocalization();
+
+            if (array_key_exists($locale, $urls)) {
+                $localizations[$locale] = $urls[$locale];
+            } elseif ($structure === null) {
+                $localizations[$locale] = '/';
+            } else {
+                $localizations[$locale] = '';
+            }
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Resolver/ParameterResolver.php]: Are `''` and `'/'` leading to different pages? I guess not, so I would omit the `elseif`.

File: src/Sulu/Bundle/WebsiteBundle/Resolver/ParameterResolver.php [hunk 7]
```diff
@@ -0,0 +1,89 @@
+<?php
+
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Resolver/ParameterResolver.php]: Missing header

File: src/Sulu/Bundle/WebsiteBundle/Resolver/ParameterResolverInterface.php [hunk 8]
```diff
@@ -0,0 +1,27 @@
+<?php
+
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Resolver/ParameterResolverInterface.php]: Missing header

File: src/Sulu/Component/Webspace/Loader/Exception/InvalidDefaultErrorTemplateException.php [hunk 9]
```diff
@@ -0,0 +1,19 @@
+<?php
+/*
+ * This file is part of the Sulu CMS.
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Loader/Exception/InvalidDefaultErrorTemplateException.php]: I think we agreed on using `This file is part of Sulu` as a header text

File: src/Sulu/Component/Webspace/Loader/Exception/InvalidDefaultErrorTemplateException.php [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Component\Webspace\Loader\Exception;
+
+/**
+ * Class InvalidDefaultErrorTemplateException
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Loader/Exception/InvalidDefaultErrorTemplateException.php]: This comment doesn't really help :-)

File: src/Sulu/Component/Webspace/Loader/Exception/InvalidDefaultErrorTemplateException.php [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+ */
+class InvalidDefaultErrorTemplateException extends WebspaceException
+{
+
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Loader/Exception/InvalidDefaultErrorTemplateException.php]: I am missing some more information here :-)

File: src/Sulu/Component/Webspace/Loader/Exception/InvalidErrorTemplateException.php [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Sulu\Component\Webspace\Loader\Exception;
+
+/**
+ * Class InvalidErrorTemplateException
+ */
+class InvalidErrorTemplateException extends WebspaceException
+{
+
+}
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Loader/Exception/InvalidErrorTemplateException.php]: Same as in previous exception: Header, comment and information.

File: src/Sulu/Component/Webspace/Loader/XmlFileLoader.php [hunk 13]
```diff
@@ -296,6 +299,39 @@ private function generateTheme()
...
+
+        foreach ($this->xpath->query('/x:webspace/x:theme/x:error-templates/x:error-template') as $errorTemplateNode) {
+            /** @var \DOMNode $errorTemplateNode */
+            $errorTemplates++;
+            if (($codeNode = $errorTemplateNode->attributes->getNamedItem('code')) !== null) {
+                $code = $codeNode->nodeValue;
+            } elseif (($defaultNode = $errorTemplateNode->attributes->getNamedItem('default')) !== null) {
+                $default = $defaultNode->nodeValue === 'true';
+                if (!$default) {
+                    throw new InvalidDefaultErrorTemplateException();
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Loader/XmlFileLoader.php]: What does this exception say at all? Is it illegal to have a node with the default value of false?

File: src/Sulu/Component/Webspace/Loader/XmlFileLoader.php [hunk 14]
```diff
@@ -296,6 +299,39 @@ private function generateTheme()
...
+                $defaultErrorTemplates++;
+                $code = 'default';
+            } else {
+                throw new InvalidErrorTemplateException();
+            }
+
+            $theme->addErrorTemplate($code, $errorTemplateNode->nodeValue);
+        }
+
+        if ($errorTemplates > 0 && $defaultErrorTemplates > 1) {
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Loader/XmlFileLoader.php]: What does this check do?

File: src/Sulu/Component/Webspace/Loader/schema/webspace/webspace-1.0.xsd [hunk 15]
```diff
@@ -111,8 +111,22 @@
...
+    <xs:complexType name="errorTemplatesType">
+        <xs:sequence>
+            <xs:element type="errorTemplateType" name="error-template" minOccurs="1" maxOccurs="unbounded"/>
+        </xs:sequence>
+    </xs:complexType>
+    <xs:complexType name="errorTemplateType">
+        <xs:simpleContent>
+            <xs:extension base="xs:string">
+                <xs:attribute name="default" type="xs:boolean"/>
+                <xs:attribute name="code" type="xs:string"/>
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Loader/schema/webspace/webspace-1.0.xsd]: Why don't you use `xs:integer` as type here?

File: src/Sulu/Component/Webspace/Loader/Exception/InvalidDefaultErrorTemplateException.php (hunk 9 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Webspace/Loader/Exception/InvalidDefaultErrorTemplateException.php]: yes i have copied from another file ... i think it is wrong in nearly all files!

File: src/Sulu/Component/Webspace/Loader/XmlFileLoader.php (hunk 13 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Webspace/Loader/XmlFileLoader.php]: yes `default = false` and no `code` defined is illegal!

File: src/Sulu/Component/Webspace/Loader/XmlFileLoader.php (hunk 14 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Webspace/Loader/XmlFileLoader.php]: if error-templates are defined only one default is legal i will add a comment

File: src/Sulu/Component/Webspace/Loader/schema/webspace/webspace-1.0.xsd (hunk 15 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Webspace/Loader/schema/webspace/webspace-1.0.xsd]: yes(=

File: src/Sulu/Component/Webspace/Loader/Exception/InvalidDefaultErrorTemplateException.php [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+ * This error represents a wrong default error template configuration
+ */
+class InvalidDefaultErrorTemplateException extends WebspaceException
+{
+    /**
+     * InvalidErrorTemplateException constructor.
+     */
+    public function __construct()
+    {
+        parent::__construct('Default cannot be false if no code is defined');
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Loader/Exception/InvalidDefaultErrorTemplateException.php]: This exception doesn't tell too much, at least some context like webspace or the template which is not correctly defined would be helpful.

File: src/Sulu/Component/Webspace/Loader/XmlFileLoader.php [hunk 17]
```diff
@@ -296,6 +299,41 @@ private function generateTheme()
...
+                $code = 'default';
+            } else {
+                throw new InvalidErrorTemplateException($template);
+            }
+
+            $theme->addErrorTemplate($code, $template);
+        }
+
+        // only one or none default error-template is legal
+        if ($errorTemplates > 0 && $defaultErrorTemplates > 1) {
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Loader/XmlFileLoader.php]: Wouldn't just the second check be enough? Would be less confusing...

Comment (wachterjohannes) [inline src/Sulu/Component/Webspace/Loader/XmlFileLoader.php]: yes i think so because if no template is defined that 0 is also legal!

**Comment by wachterjohannes** [comment]
@danrot everything fixed

File: src/Sulu/Component/Webspace/Loader/Exception/InvalidDefaultErrorTemplateException.php [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+    /**
+     * InvalidErrorTemplateException constructor.
+     * @param string $template
+     * @param int $webspace
+     */
+    public function __construct($template, $webspace)
+    {
+        parent::__construct(
+            sprintf(
+                'Default cannot of "%s" in webspace "%s" be false if no code is defined.',
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Loader/Exception/InvalidDefaultErrorTemplateException.php]: Should probably be somethiing ilke `Default of "%s" in webspace "%s" cannot be false if no code is defined.`

File: src/Sulu/Component/Webspace/Loader/Exception/InvalidErrorTemplateException.php [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+    /**
+     * InvalidErrorTemplateException constructor.
+     * @param string $template
+     * @param string $webspace
+     */
+    public function __construct($template, $webspace)
+    {
+        parent::__construct(
+            sprintf(
+                'Error template "%s" definition in "%s". It has to be defined as default or with a code.',
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Loader/Exception/InvalidErrorTemplateException.php]: `Error template "%s" in webspace "%s" has to be defined as default or with a code.`

**Pull request merged**